File: python/oris_encoding.py
#!/usr/bin/env python3
"""
ORIS Georgian Encoding Codec
============================

ORIS იყენებს single-byte custom codepage-ს, რომელიც bit-by-bit
შეესაბამება CP1251/CP1252-ის bytes 0xC0-0xE4 დიაპაზონს, მაგრამ
ეს ბაიტები რეალურად ქართულ ანბანს წარმოადგენენ.

Source: Converter_რეალური_სიმბოლოებით.xlsx (33 ქართული letter)

ASCII (0x00-0x7F) გადადის as-is.
ციფრები, სიმბოლოები, space — ჩვეულებრივი ASCII.

ეს encoding ცნობილია სხვადასხვა სახელით:
  - ORIS Unicode (პრაქტიკულად არ არის Unicode)
  - "ცრუ კირილიცა" / fake Cyrillic
  - GeorgianTransliterated

გამოყენება:
    from oris_encoding import encode_oris, decode_oris
    
    raw_bytes = encode_oris("შპს ალფა")     # → b'\\xc7\\xd0\\xd1 \\xc0\\xcb\\xce\\xc0'
    text = decode_oris(raw_bytes)            # → "შპს ალფა"
"""

# Forward mapping: Georgian Unicode char → ORIS byte
# (ანბანის თანმიმდევრობა Converter Excel-დან)
_GEO_TO_BYTE = {
    'ა': 0xC0, 'ბ': 0xC1, 'გ': 0xC2, 'დ': 0xC3, 'ე': 0xC4,
    'ვ': 0xC5, 'ზ': 0xC6, 'თ': 0xC8, 'ი': 0xC9, 'კ': 0xCA,
    'ლ': 0xCB, 'მ': 0xCC, 'ნ': 0xCD, 'ო': 0xCF, 'პ': 0xD0,
    'ჟ': 0xD1, 'რ': 0xD2, 'ს': 0xD3, 'ტ': 0xD4, 'უ': 0xD6,
    'ფ': 0xD7, 'ქ': 0xD8, 'ღ': 0xD9, 'ყ': 0xDA, 'შ': 0xDB,
    'ჩ': 0xDC, 'ც': 0xDD, 'ძ': 0xDE, 'წ': 0xDF, 'ჭ': 0xE0,
    'ხ': 0xE1, 'ჯ': 0xE3, 'ჰ': 0xE4,
}

# Reverse mapping: ORIS byte → Georgian Unicode char
_BYTE_TO_GEO = {v: k for k, v in _GEO_TO_BYTE.items()}


def encode_oris(text: str, errors: str = 'replace') -> bytes:
    """
    ქართული Unicode ტექსტი → ORIS bytes.
    
    Args:
        text:   Python string (UTF-8 internal)
        errors: 'replace' (default) — non-mapped chars become '?'
                'strict'             — raise ValueError
                'ignore'             — skip non-mapped chars
    
    Returns:
        bytes — ORIS-compatible single-byte sequence
    """
    out = bytearray()
    for ch in text:
        if ch in _GEO_TO_BYTE:
            out.append(_GEO_TO_BYTE[ch])
        elif ord(ch) < 0x80:
            # ASCII pass-through
            out.append(ord(ch))
        else:
            if errors == 'strict':
                raise ValueError(f"Char {ch!r} (U+{ord(ch):04X}) not in ORIS encoding")
            elif errors == 'replace':
                out.append(ord('?'))
            # 'ignore' — skip
    return bytes(out)


def decode_oris(data: bytes, errors: str = 'replace') -> str:
    """
    ORIS bytes → ქართული Unicode ტექსტი.
    """
    out = []
    for b in data:
        if b in _BYTE_TO_GEO:
            out.append(_BYTE_TO_GEO[b])
        elif b < 0x80:
            out.append(chr(b))
        else:
            if errors == 'strict':
                raise ValueError(f"Byte 0x{b:02X} not in ORIS encoding")
            elif errors == 'replace':
                out.append('?')
            # 'ignore' — skip
    return ''.join(out)


def to_oris_field(text: str, length: int, pad_byte: int = 0x20) -> bytes:
    """
    String → fixed-length ORIS field (Clarion STRING type).
    
    შემთხვევები:
    - text უფრო მოკლე: pad-ი space-ით (0x20)
    - text უფრო გრძელი: truncate-ი length-ზე
    
    გამოყენება Clarion's FIELD_STRING (type 0x12)-ისთვის.
    """
    encoded = encode_oris(text)
    if len(encoded) >= length:
        return encoded[:length]
    return encoded + bytes([pad_byte]) * (length - len(encoded))


def from_oris_field(data: bytes) -> str:
    """
    Fixed-length ORIS field → trimmed Unicode string.
    """
    return decode_oris(data).rstrip()


# ===================================================================
# Self-test
# ===================================================================

if __name__ == '__main__':
    print("=== ORIS Encoding Codec Test ===\n")
    
    # ტესტი 1: encode/decode round-trip
    samples = [
        "შპს ალფა",
        "ნედლეული მასალები",
        "გაყიდვები",
        "ABC 123 ქართული",
        "ვალდებულებანი მიმწოდებლების მიმართ",
    ]
    
    for s in samples:
        encoded = encode_oris(s)
        decoded = decode_oris(encoded)
        status = "✓" if decoded == s else "✗"
        hex_str = ' '.join(f'{b:02x}' for b in encoded[:30])
        print(f"  {status} {s!r}")
        print(f"     bytes ({len(encoded)}): {hex_str}{'...' if len(encoded) > 30 else ''}")
        print(f"     decoded: {decoded!r}")
        print()
    
    # ტესტი 2: fixed-length field
    print("=== Fixed-length field test ===")
    name = "შპს ალფა"
    field = to_oris_field(name, 30)
    print(f"  Input:    {name!r}")
    print(f"  Field:    {field!r}")
    print(f"  Length:   {len(field)} bytes")
    print(f"  Restored: {from_oris_field(field)!r}")
    print()
    
    # ტესტი 3: სრულდება როგორ ჩანდა ORIS-ში თუ ფაილს ჩავწერთ ბაიტურად
    # და გავხსნით CP1251 viewer-ით
    print("=== How it appears in CP1251 viewer ===")
    s = "შპს ალფა"
    raw = encode_oris(s)
    print(f"  Unicode source: {s!r}")
    print(f"  Raw bytes:      {raw.hex(' ')}")
    print(f"  CP1251 render:  {raw.decode('cp1251', errors='replace')!r}")
    print(f"  CP1252 render:  {raw.decode('cp1252', errors='replace')!r}")
    print()
    print("  ↑ ეს არის ის, რასაც ხედავ ORIS-ში 'რუსული ფონტით'.")

File: csharp/OrisTpsWriter.Core/OrisEncoding.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace OrisTpsWriter.Core
{
    /// <summary>
    /// ORIS Georgian Encoding Codec.
    ///
    /// ORIS იყენებს single-byte custom codepage-ს, რომელიც ბაიტებს 0xC0-0xE4
    /// ქართულ ანბანს უსაბამებს. რენდერდება როგორც "А Б В" რუსული ფონტით
    /// ან "À Á Â" ლათინური ფონტით — ფაილში ბაიტი ერთი და იგივეა.
    ///
    /// წყარო: Converter_რეალური_სიმბოლოებით.xlsx (33 ქართული ასო).
    /// დადასტურებულია რეალური ARN.TPS ფაილით (ოქრიაშვილი, გოგიჩაძე და ა.შ.).
    /// </summary>
    public static class OrisEncoding
    {
        // Georgian Unicode char → ORIS byte
        private static readonly Dictionary<char, byte> GeoToByte = new()
        {
            {'\u10D0', 0xC0}, // ა
            {'\u10D1', 0xC1}, // ბ
            {'\u10D2', 0xC2}, // გ
            {'\u10D3', 0xC3}, // დ
            {'\u10D4', 0xC4}, // ე
            {'\u10D5', 0xC5}, // ვ
            {'\u10D6', 0xC6}, // ზ
            {'\u10D7', 0xC8}, // თ
            {'\u10D8', 0xC9}, // ი
            {'\u10D9', 0xCA}, // კ
            {'\u10DA', 0xCB}, // ლ
            {'\u10DB', 0xCC}, // მ
            {'\u10DC', 0xCD}, // ნ
            {'\u10DD', 0xCF}, // ო
            {'\u10DE', 0xD0}, // პ
            {'\u10DF', 0xD1}, // ჟ
            {'\u10E0', 0xD2}, // რ
            {'\u10E1', 0xD3}, // ს
            {'\u10E2', 0xD4}, // ტ
            {'\u10E3', 0xD6}, // უ
            {'\u10E4', 0xD7}, // ფ
            {'\u10E5', 0xD8}, // ქ
            {'\u10E6', 0xD9}, // ღ
            {'\u10E7', 0xDA}, // ყ
            {'\u10E8', 0xDB}, // შ
            {'\u10E9', 0xDC}, // ჩ
            {'\u10EA', 0xDD}, // ც
            {'\u10EB', 0xDE}, // ძ
            {'\u10EC', 0xDF}, // წ
            {'\u10ED', 0xE0}, // ჭ
            {'\u10EE', 0xE1}, // ხ
            {'\u10EF', 0xE3}, // ჯ
            {'\u10F0', 0xE4}, // ჰ
        };

        // ORIS byte → Georgian Unicode char (reverse map)
        private static readonly Dictionary<byte, char> ByteToGeo;

        static OrisEncoding()
        {
            ByteToGeo = new Dictionary<byte, char>();
            foreach (var kvp in GeoToByte)
                ByteToGeo[kvp.Value] = kvp.Key;
        }

        /// <summary>
        /// ქართული/ASCII string → ORIS bytes.
        /// ASCII (0x00-0x7F) გადადის as-is. Non-mapped chars → '?'.
        /// </summary>
        public static byte[] Encode(string text)
        {
            if (string.IsNullOrEmpty(text))
                return Array.Empty<byte>();

            var outBytes = new List<byte>(text.Length);
            foreach (char ch in text)
            {
                if (GeoToByte.TryGetValue(ch, out byte b))
                    outBytes.Add(b);
                else if (ch < 0x80)
                    outBytes.Add((byte)ch);
                else
                    outBytes.Add((byte)'?'); // non-mapped
            }
            return outBytes.ToArray();
        }

        /// <summary>
        /// ORIS bytes → ქართული/ASCII string.
        /// </summary>
        public static string Decode(byte[] data)
        {
            if (data == null || data.Length == 0)
                return string.Empty;

            var sb = new StringBuilder(data.Length);
            foreach (byte b in data)
            {
                if (ByteToGeo.TryGetValue(b, out char ch))
                    sb.Append(ch);
                else if (b < 0x80)
                    sb.Append((char)b);
                else
                    sb.Append('?');
            }
            return sb.ToString();
        }

        /// <summary>
        /// String → fixed-length ORIS field (Clarion STRING type).
        /// უფრო მოკლე → space-padded (0x20). უფრო გრძელი → truncated.
        /// </summary>
        public static byte[] ToFixedField(string text, int length, byte padByte = 0x20)
        {
            byte[] encoded = Encode(text ?? string.Empty);
            byte[] result = new byte[length];

            // Fill with pad byte first
            for (int i = 0; i < length; i++)
                result[i] = padByte;

            // Copy encoded bytes (truncate if too long)
            int copyLen = Math.Min(encoded.Length, length);
            Array.Copy(encoded, result, copyLen);

            return result;
        }

        /// <summary>
        /// Fixed-length ORIS field → trimmed string.
        /// </summary>
        public static string FromFixedField(byte[] data)
        {
            return Decode(data).TrimEnd();
        }
    }
}
